platforms/android/tangram/src/main/java/com/mapzen/tangram/MapController.java

```diff
@@ -1361,45 +1359,68 @@ public void onSurfaceCreated(final GL10 gl, final EGLConfig config) {
+            synchronized(this) {
+                request = httpRequestHandles.get(requestHandle);
+            }
+            if (request != null) {
+                httpHandler.cancelRequest(request);
+            }
+            syncRemoveHttpHandle(requestHandle);
         }
-        httpHandler.onCancel(requestHandle);
     }
 
+
+
     @Keep
     void startUrlRequest(@NonNull final String url, final long requestHandle) {
         if (httpHandler == null) {
             return;
         }
 
-        final Callback callback = new Callback() {
+        final HttpHandler.Callback callback = new HttpHandler.Callback() {
             @Override
-            public void onFailure(final Call call, final IOException e) {
-                nativeOnUrlComplete(mapPointer, requestHandle, null, e.getMessage());
+            public void onFailure(@Nullable final IOException e) {
+                String msg = (e == null) ? null : e.getMessage();
+                nativeOnUrlComplete(mapPointer, requestHandle, null, msg);
+                syncRemoveHttpHandle(requestHandle);
             }
 
             @Override
-            public void onResponse(final Call call, final Response response) throws IOException {
-                if (!response.isSuccessful()) {
-                    nativeOnUrlComplete(mapPointer, requestHandle, null, response.message());
-                    throw new IOException("Unexpected response code: " + response + " for URL: " + url);
-                }
-                final ResponseBody body = response.body();
-                if (body == null) {
-                    nativeOnUrlComplete(mapPointer, requestHandle, null, response.message());
-                    throw new IOException("Unexpected null body for URL: " + url);
+            public void onResponse(final int code, @Nullable final byte[] rawDataBytes) {
+                // TODO: Use of returned error code and headers for better network response logging/retries, etc
+                if (code >= 200 && code < 300) {
+                    nativeOnUrlComplete(mapPointer, requestHandle, rawDataBytes, null);
                 }
                 else {
-                    final byte[] bytes = body.bytes();
-                    nativeOnUrlComplete(mapPointer, requestHandle, bytes, null);
+                    nativeOnUrlComplete(mapPointer, requestHandle, null, "Unexpected response code: " + code + " for URL: " + url);
                 }
+                syncRemoveHttpHandle(requestHandle);
+            }
+
+            @Override
+            public void onCancel() {
+                nativeOnUrlComplete(mapPointer, requestHandle, null, null);
+                syncRemoveHttpHandle(requestHandle);
             }
         };
 
-        httpHandler.onRequest(url, callback, requestHandle);
+        Object request = httpHandler.startRequest(url, callback);
+        if (request != null) {
+            synchronized (this) {
+                httpRequestHandles.put(requestHandle, request);
```

Comment:
I think there's a subtle edge case to be careful about here. If `startRequest` completes immediately (like in `DefaultHttpHandler` when the URL can't be parsed) then the callback is executed before the request is added to the handle map. I think this the worst case is that we get some request entries in the map that never get removed, which ends not being a problem with `DefaultHttpHandler`. It might not be worth guarding against this now, but we should keep aware of it.

Reply:
Good Point.
For `DefaultHttpHandler` purposes its not a problem, as the returned `request` object is `null` and won't be added to the map. But there could be cases where callbacks gets executed before returned `request` object is put in the map, which I think is kinda not trivial to solve. Will leave it for subsequent PR.